File: app/src/main/java/com/jcminarro/authexample/internal/network/ApiClient.java
package com.jcminarro.authexample.internal.network;

import java.io.IOException;

import retrofit2.Call;
import retrofit2.Response;

public class ApiClient<T> {

    protected final T endpoint;

    public ApiClient(T endpoint) {
        this.endpoint = endpoint;
    }

    protected <U> U evaluateCall(Call<U> call) throws IOException {
        Response<U> response = call.execute();
        if (!response.isSuccessful()) {
            throw new APIIOException(response.raw());
        }
        return response.body();
    }
}

File: app/src/main/java/com/jcminarro/authexample/internal/network/ApiClient.kt
package com.jcminarro.authexample.internal.network

import arrow.core.Try
import retrofit2.Call
import java.io.IOException

open class ApiClient<T>(protected val endpoint: T) {

    @Throws(IOException::class)
    protected fun <U> evaluateCall(call: Call<U>): U {
        val response = call.execute()
        if (!response.isSuccessful) {
            throw APIIOException(response.raw())
        }
        return response.body()
    }

    protected fun <U> evaluateCallFP(call: Call<U>): Try<U> = Try {evaluateCall(call)}
}
